### PR TITLE
Python3 cleanup

### PR DESCRIPTION
**Pull request by dannon**
Now that we're officially py3-only, we can clean up the cruft and standardize a few things.

I used some upgrade tooling to do stuff here like swap to a single class style, drop `six` compatibility shims in some places where it was straightforward, standardize print formatting and drop __future__ imports, etc.  There's a good bit more to do, some tweaky cases that we'll want to handle manually, but I wanted to draft this and get tests running.

### CODE REVIEW

File: lib/galaxy/datatypes/converters/bed_to_gff_converter.py [hunk 1]
```diff
@@ -1,6 +1,5 @@
 #!/usr/bin/env python
 # This code exists in 2 places: ~/datatypes/converters and ~/tools/filters
-from __future__ import print_function
```
Comment (dannon) [inline lib/galaxy/datatypes/converters/bed_to_gff_converter.py:3]: I might need to back out tweaks like this, and other converters, that can/should still run with an older python?

**Comment by nuwang** [comment]
What tooling did you find useful for this?  Would help with porting more things to py3. Happy to say bye to py2 at the earliest possible :-)

File: lib/galaxy/datatypes/converters/bed_to_gff_converter.py (hunk 1 above)
Comment (nsoranzo) [reply lib/galaxy/datatypes/converters/bed_to_gff_converter.py:3]: The tool XML already has `<requirement type="package" version="3.7">python</requirement>`, so I think it's fine.

**Comment by dannon** [comment]
@nuwang a whole lot of `codemod` and pyupgrade

File: lib/galaxy/datatypes/converters/maf_to_fasta_converter.py [hunk 2]
```diff
@@ -16,7 +15,7 @@ def __main__():
     input_name = sys.argv.pop(1)
     count = 0
     with open(output_name, 'w') as out:
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/converters/maf_to_fasta_converter.py]:
```suggestion
    with open(output_name, 'w') as out, open(input_name) as in:
```

File: lib/galaxy/datatypes/converters/maf_to_fasta_converter.py [hunk 3]
```diff
@@ -16,7 +15,7 @@ def __main__():
...
-        for count, block in enumerate(bx.align.maf.Reader(open(input_name, 'r'))):
+        for count, block in enumerate(bx.align.maf.Reader(open(input_name))):
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/converters/maf_to_fasta_converter.py]:
```suggestion
        for count, block in enumerate(bx.align.maf.Reader(in)):
```

File: lib/galaxy/datatypes/data.py [hunk 4]
```diff
@@ -488,7 +483,7 @@ def display_as_markdown(self, dataset_instance, markdown_format_helpers):
         if self.is_binary:
             result = "*cannot display binary content*\n"
         else:
-            contents = open(dataset_instance.file_name, "r").read(DEFAULT_MAX_PEEK_SIZE)
+            contents = open(dataset_instance.file_name).read(DEFAULT_MAX_PEEK_SIZE)
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/data.py]:
```suggestion
            with open(dataset_instance.file_name) as f:
                contents = f.read(DEFAULT_MAX_PEEK_SIZE)
```

File: lib/galaxy/datatypes/data.py [hunk 5]
```diff
@@ -506,7 +501,7 @@ def _yield_user_file_content(self, trans, from_dataset, filename):
             # This is returning to the browser, it needs to be encoded.
             # TODO Ideally this happens a layer higher, but this is a bad
             # issue affecting many tools
-            return sanitize_html(open(filename, 'r').read()).encode('utf-8')
+            return sanitize_html(open(filename).read()).encode('utf-8')
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/data.py]:
```suggestion
            with open(filename) as f:
                return sanitize_html(f.read()).encode('utf-8')
```

File: lib/galaxy/datatypes/display_applications/parameters.py [hunk 6]
```diff
@@ -190,7 +190,7 @@ def url(self):
         base_url = self.trans.request.base
         if self.parameter.strip_https and base_url[: 5].lower() == 'https':
             base_url = "http%s" % base_url[5:]
-        return "%s%s" % (base_url,
+        return "{}{}".format(base_url,
                          self.trans.app.url_for(controller='dataset',
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/display_applications/parameters.py]: Indentation seems wrong here.

File: lib/galaxy/datatypes/genetics.py [hunk 7]
```diff
@@ -712,7 +712,7 @@ def set_peek(self, dataset, **kwd):
         if not dataset.dataset.purged:
             pp = os.path.join(dataset.extra_files_path, '%s.pheno' % dataset.metadata.base_name)
             try:
-                p = open(pp, 'r').readlines()
+                p = open(pp).readlines()
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/genetics.py]:
```suggestion
                with open(pp) as f:
                    p = f.readlines()
```

File: lib/galaxy/datatypes/genetics.py [hunk 8]
```diff
@@ -727,7 +727,7 @@ def get_peek(self, dataset):
         """
         pp = os.path.join(dataset.extra_files_path, '%s.pheno' % dataset.metadata.base_name)
         try:
-            p = open(pp, 'r').readlines()
+            p = open(pp).readlines()
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/genetics.py]:
```suggestion
            with open(pp) as f:
                p = f.readlines()
```

File: lib/galaxy/datatypes/genetics.py [hunk 9]
```diff
@@ -738,7 +738,7 @@ def get_file_peek(self, filename):
         """
         h = '## rexpression get_file_peek: no file found'
         try:
-            h = open(filename, 'r').readlines()
+            h = open(filename).readlines()
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/genetics.py]:
```suggestion
            with open(filename) as f:
                h = f.readlines()
```

File: lib/galaxy/datatypes/genetics.py [hunk 10]
```diff
@@ -788,7 +788,7 @@ def set_meta(self, dataset, **kwd):
         pp = os.path.join(dataset.extra_files_path, pn)
         dataset.metadata.pheno_path = pp
         try:
-            pf = open(pp, 'r').readlines()  # read the basename.phenodata in the extra_files_path
+            pf = open(pp).readlines()  # read the basename.phenodata in the extra_files_path
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/genetics.py]:
```suggestion
            with open(pp) as f:
                pf = f.readlines()  # read the basename.phenodata in the extra_files_path
```

File: lib/galaxy/datatypes/sequence.py [hunk 11]
```diff
@@ -780,7 +778,7 @@ def process_split_file(data):
         sequence_count = long(args['num_sequences'])
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/sequence.py]:
```suggestion
        sequence_count = int(args['num_sequences'])
```

File: lib/galaxy/datatypes/sequence.py [hunk 12]
```diff
@@ -36,8 +35,7 @@
 from galaxy.util.image_util import check_image_type
 from . import data
 
-if sys.version_info > (3,):
-    long = int
+long = int
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/sequence.py]: Should we replace `long` with `int` in the file instead?

File: lib/galaxy/datatypes/sequence.py [hunk 13]
```diff
@@ -780,7 +778,7 @@ def process_split_file(data):
         sequence_count = long(args['num_sequences'])
 
         if 'toc_file' in args:
-            toc_file = json.load(open(args['toc_file'], 'r'))
+            toc_file = json.load(open(args['toc_file']))
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/sequence.py]:
```suggestion
            with open(args['toc_file']) as f:
                toc_file = json.load(f)
```

File: lib/galaxy/datatypes/sniff.py [hunk 14]
```diff
@@ -13,16 +12,9 @@
 import shutil
 import sys
 import tempfile
+import urllib
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/sniff.py]:
```suggestion
import urllib.request
```

File: lib/galaxy/datatypes/tabular.py [hunk 15]
```diff
@@ -30,8 +28,7 @@
 from galaxy.util import compression_utils
 from . import dataproviders
 
-if sys.version_info > (3,):
-    long = int
+long = int
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/tabular.py]: Should we replace `long` with `int` in the file instead?

File: lib/galaxy/datatypes/tabular.py [hunk 16]
```diff
@@ -125,7 +122,7 @@ def display_data(self, trans, dataset, preview=False, filename=None, to_ext=None
                                        column_types=column_types)
 
     def display_as_markdown(self, dataset_instance, markdown_format_helpers):
-        contents = open(dataset_instance.file_name, "r").read(data.DEFAULT_MAX_PEEK_SIZE)
+        contents = open(dataset_instance.file_name).read(data.DEFAULT_MAX_PEEK_SIZE)
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/tabular.py]:
```suggestion
        with open(dataset_instance.file_name) as f:
            contents = f.read(data.DEFAULT_MAX_PEEK_SIZE)
```

File: lib/galaxy/datatypes/tabular.py [hunk 17]
```diff
@@ -1002,7 +998,7 @@ def sniff(self, filename):
         """ Return True if if recognizes dialect and header. """
         try:
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/tabular.py]: We can get rid of `try:` and the `except` block.

File: lib/galaxy/datatypes/tabular.py [hunk 18]
```diff
@@ -1002,7 +998,7 @@ def sniff(self, filename):
...
             # check the dialect works
-            reader = csv.reader(open(filename, 'r'), self.dialect)
+            reader = csv.reader(open(filename), self.dialect)
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/tabular.py]:
```suggestion
            with open(filename) as f:
                reader = csv.reader(f, self.dialect)
```

File: lib/galaxy/datatypes/tabular.py [hunk 19]
```diff
@@ -1031,7 +1027,7 @@ def sniff(self, filename):
                     pass
 
             # Optional: Check Python's csv comes up with a similar dialect
-            auto_dialect = csv.Sniffer().sniff(open(filename, 'r').read(self.big_peek_size))
+            auto_dialect = csv.Sniffer().sniff(open(filename).read(self.big_peek_size))
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/tabular.py]:
```suggestion
            with open(filename) as f:
                big_peek = f.read(self.big_peek_size)
            auto_dialect = csv.Sniffer().sniff(big_peek)
```

File: lib/galaxy/datatypes/tabular.py [hunk 20]
```diff
@@ -1046,7 +1042,7 @@ def sniff(self, filename):
             Note Without checking the dialect returned by sniff
                   this test may be checking the wrong dialect.
             """
-            if not csv.Sniffer().has_header(open(filename, 'r').read(self.big_peek_size)):
+            if not csv.Sniffer().has_header(open(filename).read(self.big_peek_size)):
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/tabular.py]:
```suggestion
            if not csv.Sniffer().has_header(big_peek):
```

**Review by nsoranzo — COMMENTED** [review]
Reviewed around a 10%, I will be on holiday from tomorrow so submitting what I have.

**Comment by dannon** [comment]
@nsoranzo I very much appreciate the eyes on this, but can we look at this with an eye towards whether or not the changes are correct, and follow up with other suggested changes later?  It goes stale on just about every commit to dev :/

File: lib/galaxy/datatypes/sequence.py (hunk 12 above)
Comment (dannon) [reply lib/galaxy/datatypes/sequence.py]: I went ahead and did this across the codebase; good idea.

**Comment by nsoranzo** [comment]
I reviewed from top till lib/galaxy/jobs/transfer_manager.py if any one wants to go ahead while I'm off.

**Comment by dannon** [comment]
Looks like the extra changes might have broken two tests?  (they did all pass green before -- I'll rerun and then dig deeper if they fail again).

**Comment by nsoranzo** [comment]
> Looks like the extra changes might have broken two tests? (they did all pass green before -- I'll rerun and then dig deeper if they fail again).

Yep, sorry about that, I think I've solved the issue now.

**Comment by dannon** [comment]
> 
> 
> > Looks like the extra changes might have broken two tests? (they did all pass green before -- I'll rerun and then dig deeper if they fail again).
> 
> Yep, sorry about that, I think I've solved the issue now.

No worries, thanks for fixing it up.  And enjoy your holiday!

File: lib/galaxy/managers/citations.py [hunk 21]
```diff
@@ -146,10 +146,10 @@ def to_bibtex(self):
                 log.exception("Failed to fetch bibtex for DOI %s", self.__doi)
 
         if self.raw_bibtex is DoiCitation.BIBTEX_UNSET:
-            return """@MISC{%s,
-                DOI = {%s},
-                note = {Failed to fetch BibTeX for DOI.}
-            }""" % (self.__doi, self.__doi)
+            return """@MISC{{{},
```
Comment (bgruening) [inline lib/galaxy/managers/citations.py]: This is not more readable than before. Its worse, hopefully we can soon use f-strings.

File: lib/galaxy/managers/history_contents.py [hunk 22]
```diff
@@ -397,7 +397,7 @@ def _contained_id_map(self, id_list):
             .options(eagerload('dataset.actions'))
             .options(eagerload('tags'))
             .options(eagerload('annotations')))
-        return dict((row.id, row) for row in query.all())
+        return {row.id: row for row in query.all()}
```
Comment (bgruening) [inline lib/galaxy/managers/history_contents.py:400]: I would prefer dict() and it is supposed to be faster than {} ... but I guess we do not care much?

**Review by bgruening — COMMENTED** [review]
review included up until  lib/galaxy/model/item_attrs.py

File: lib/galaxy/managers/history_contents.py (hunk 22 above)
Comment (dannon) [reply lib/galaxy/managers/history_contents.py:400]: Weird, I saw the comprehension format as faster in a quick test.  What situations is it not supposed to be good for?

```
[yoplait@incus:~/work/galaxy] python3 -m timeit -s 'import string' 'b = {i: 1 for i in string.digits}'
500000 loops, best of 5: 617 nsec per loop
[yoplait@incus:~/work/galaxy] python3 -m timeit -s 'import string' 'b = dict((i,1) for i in string.digits)'
200000 loops, best of 5: 1.33 usec per loop
```

File: lib/galaxy/managers/citations.py (hunk 21 above)
Comment (dannon) [reply lib/galaxy/managers/citations.py]: Yeah, this particular one is maybe worse with all the other braces involved.   Want me to update it to have the variable names inside?  That might help here.  (done in https://github.com/galaxyproject/galaxy/pull/9923/commits/a2d1ec58040c9ee2eb296ab7d03f8c8415c40a29)

I actually *did* do f-string conversion at one point, so the good news is that it's very easy to do when we're ready.  The problem is that it's 3.6+ and I think we support 3.5.

Comment (bgruening) [reply lib/galaxy/managers/citations.py]: Jupp, we need at first to drop 3.5. So we should wait. We need to fix uwsgi-support for 3.7, 3.8+ so we can also use those python version :(

> Want me to update it to have the variable names inside?

I think that would be nice.

File: lib/galaxy/managers/history_contents.py (hunk 22 above)
Comment (bgruening) [reply lib/galaxy/managers/history_contents.py:400]: Mh, ok, interesting. Lets leave it like that. Thanks.

File: lib/galaxy/util/object_wrapper.py [hunk 23]
```diff
@@ -108,10 +106,11 @@
 
 INVALID_CHARACTER = "X"
 
-if sys.version_info > (3, 0):
-    # __coerce__ doesn't do anything under Python anyway.
-    def coerce(x, y):
-        return x
+# __coerce__ doesn't do anything under Python anyway.
+
```
Comment (bgruening) [inline lib/galaxy/util/object_wrapper.py]: empty lines could be removed or removed completely?

File: lib/galaxy/util/rules_dsl.py [hunk 24]
```diff
@@ -73,7 +70,7 @@ class AddColumnMetadataRuleDefinition(BaseRuleDefinition):
     rule_type = "add_column_metadata"
 
     def validate_rule(self, rule):
-        _ensure_rule_contains_keys(rule, {"value": six.string_types})
+        _ensure_rule_contains_keys(rule, {"value": (str,)})
```
Comment (bgruening) [inline lib/galaxy/util/rules_dsl.py]: can this be `str`?

**Review by bgruening — APPROVED** [review]
Ok, I looked at all files starting with @nsoranzo ones. Just two small questions but please feel free to merge. Those rebeases are probably annoying.

And we can remove six from all our requirements files.

File: lib/galaxy/util/rules_dsl.py (hunk 24 above)
Comment (dannon) [reply lib/galaxy/util/rules_dsl.py]: Yep, sure can.  I'll swap it.

File: lib/galaxy/util/object_wrapper.py (hunk 23 above)
Comment (dannon) [reply lib/galaxy/util/object_wrapper.py]: Yeah, I put the comment inside the function, thank you!

**Comment by dannon** [comment]
The one failing selenium test passes locally:  
image

**Comment by bgruening** [comment]
Thanks a lot @dannon and @nsoranzo!

**Pull request merged**